fw/src/app/belts.cpp

```diff
@@ -0,0 +1,174 @@
+
+/*******************************************************************************
+*                                I N C L U D E S                               *
+*******************************************************************************/ 
+
+#include "belts.h"
+#include "scheduler.h"
+
+/*******************************************************************************
+*                               C O N S T A N T S                              *
+*******************************************************************************/ 
+
+#define BELTS_NAV_RATE 750
+#define BELTS_STARTING_RATE 100
+
+/*******************************************************************************
+*                      D A T A    D E C L A R A T I O N S                      *
+*******************************************************************************/ 
+
+typedef struct
+{
+    struct pt thread;
+    belts_state_E state;
+    belts_state_E des_state;
+    uint16_t top_belt_steps;
+    uint16_t bottom_belt_steps;
+} belts_data_S;
+
+/*******************************************************************************
+*          P R I V A T E    F U N C T I O N    D E C L A R A T I O N S         *
+*******************************************************************************/
+
+static belts_state_E belts_update_state(belts_state_E curr_state);
+
+/*******************************************************************************
+*                 S T A T I C    D A T A    D E F I N I T I O N S              *
+*******************************************************************************/ 
+
+static belts_data_S belts_data = 
+{
+    .state = BELTS_STATE_IDLE,
+    .des_state = BELTS_STATE_IDLE,
+    .top_belt_steps = 0,
+    .bottom_belt_steps = 0,
+
+};
+
+/*******************************************************************************
+*                      P R I V A T E    F U N C T I O N S                      *
+*******************************************************************************/
+
+static belts_state_E belts_parseState(char* s)
+{
+    belts_state_E state = BELTS_STATE_COUNT;
+    if (strcmp(s, "idle") == 0)
+    {
+        state = BELTS_STATE_IDLE;
+    }
+    else if (strcmp(s, "active") == 0)
+    {
+        state = BELTS_STATE_ACTIVE;
+    }
+    else
+    {
+        // do nothing
+    }
+    return state;
+}
+
+static belts_state_E belts_update_state(belts_state_E curr_state)
+{
+    belts_state_E next_state = curr_state;
+    bool top = false;
+    bool bottom = false;
+
+    switch (curr_state)
+    {
+        case BELTS_STATE_IDLE:
+            if (belts_data.des_state == BELTS_STATE_ACTIVE)
+            {
+                // only go to active once per write from SW to des_state
+                belts_data.des_state = BELTS_STATE_IDLE;
```

Comment:
This gets overwritten by `belts_core_comms_setDesState()`. Since core_comms spams msgs, it'll keep trying to set the belts to active after belt movement stops. 

We could make SW side send a single "active" msg and all subsequent messages "idle", because when `belts.cpp` is active it doesn't care about desired state.

Reply:
Could we have a firmware flag that only flips when the software has commanded idle state after commanding active?